src/Microsoft.DotNet.XHarness.Android/AdbRunner.cs

```diff
@@ -413,8 +413,38 @@ public List<string> PullFiles(string devicePath, string localPath)
         }
 
 
+
         public string? GetDeviceToUse(ILogger logger, string apkRequiredProperty, string propertyName)
         {
+            var allDevicesAndTheirProperties = GetAllDevicesToUse(logger, apkRequiredProperty, propertyName);
+            if (allDevicesAndTheirProperties.Count > 0)
+            {
+                var firstAvailableCompatible = allDevicesAndTheirProperties.First();
+                logger.LogDebug($"Using first-found compatible device of {allDevicesAndTheirProperties.Count} total- serial: '{firstAvailableCompatible.Key}' - {propertyName}: {firstAvailableCompatible.Value}");
+                return firstAvailableCompatible.Key;
+            }
+            return null;
+        }
+
+        public string? GetUniqueDeviceToUse(ILogger logger, string apkRequiredProperty, string propertyName)
+        {
+            var devices = GetAllDevicesToUse(logger, apkRequiredProperty, propertyName);
+            if (devices.Count == 0)
+            {
+                logger.LogError("Cannot find a device with app installed, please check that a device is attached or run install first");
+                return null;
+            }
+            else if (devices.Count > 1)
+            {
+                logger.LogError("There are more than one device with the app installed, please provide --device-id to choose the required one");
+                return null;
+            }
+            return devices.Keys.First();
+        }
+
+        public Dictionary<string, string> GetAllDevicesToUse(ILogger logger, string apkRequiredProperty, string propertyName)
+        {
+
             Dictionary<string, string?> allDevicesAndTheirProperties = new Dictionary<string, string?>();

```

Comment:
I think there's a mismatch - `result` is `Dictionary<string, string?>` but the return type is `Dictionary<string, string>`.

Reply:
I think maybe you could use a different variable when you re-create the `string, string` version in the `ToDictionary`

Reply:
Actually I filter out all null values, so I'm going to convert it to `Dictionary<string, string>` explicitely